File: backend/engine/dfa_engine.py
try:
    from models.schemas import DFAMachine, SimulationResult, SimulationStepTrace
except ImportError:
    from ..models.schemas import DFAMachine, SimulationResult, SimulationStepTrace

def simulate_dfa(machine: DFAMachine, input_str: str) -> SimulationResult:
    traces: List[SimulationStepTrace] = []
    current_state = machine.startState
    chars = list(input_str)

    state_lookup = {s.id: s.label for s in machine.states}

    # Initial Step
    is_initial_accept = current_state in machine.acceptStates
    traces.append(
        SimulationStepTrace(
            stepIndex=0,
            activeStates=[current_state],
            remainingInput=input_str,
            readSymbol=None,
            activeTransitionId=None,
            actionSummary=f"Started at initial state '{state_lookup.get(current_state, current_state)}'",
            status="RUNNING" if len(chars) > 0 else ("ACCEPTED" if is_initial_accept else "REJECTED"),
        )
    )

    if len(chars) == 0:
        return SimulationResult(
            accepted=is_initial_accept,
            finalStatus="ACCEPTED" if is_initial_accept else "REJECTED",
            totalSteps=1,
            traces=traces,
            message=f"Empty string ε {'accepted' if is_initial_accept else 'rejected'}.",
        )

    for i, sym in enumerate(chars):
        # Find matching transition
        transition = next(
            (t for t in machine.transitions if t.from_state == current_state and t.symbol == sym),
            None
        )

        remaining = "".join(chars[i+1:])

        if transition is None:
            traces.append(
                SimulationStepTrace(
                    stepIndex=len(traces),
                    activeStates=[],
                    remainingInput=remaining,
                    readSymbol=sym,
                    activeTransitionId=None,
                    actionSummary=f"No transition defined from '{state_lookup.get(current_state, current_state)}' on symbol '{sym}'. Trap/Dead state entered.",
                    status="REJECTED",
                )
            )
            return SimulationResult(
                accepted=False,
                finalStatus="REJECTED",
                totalSteps=len(traces),
                traces=traces,
                message=f"Rejected: No transition for symbol '{sym}' from state '{state_lookup.get(current_state, current_state)}'.",
            )

        current_state = transition.to_state
        is_last = (i == len(chars) - 1)
        is_accept = current_state in machine.acceptStates

        traces.append(
            SimulationStepTrace(
                stepIndex=len(traces),
                activeStates=[current_state],
                remainingInput=remaining,
                readSymbol=sym,
                activeTransitionId=transition.id,
                actionSummary=f"Read '{sym}' -> Transitioned to '{state_lookup.get(current_state, current_state)}'",
                status=("ACCEPTED" if is_accept else "REJECTED") if is_last else "RUNNING",
            )
        )

    accepted = current_state in machine.acceptStates
    return SimulationResult(
        accepted=accepted,
        finalStatus="ACCEPTED" if accepted else "REJECTED",
        totalSteps=len(traces),
        traces=traces,
        message=f"Simulation finished. Ended in {'accept' if accepted else 'non-accept'} state '{state_lookup.get(current_state, current_state)}'.",
    )

File: frontend/src/engine/dfaEngine.ts
import { DFAMachine, SimulationResult, SimulationStepTrace } from '../types/automata';

export function simulateDFA(machine: DFAMachine, inputString: string): SimulationResult {
  const traces: SimulationStepTrace[] = [];
  let currentState = machine.startState;
  
  // Initial Step 0
  traces.push({
    stepIndex: 0,
    activeStates: [currentState],
    currentSymbol: inputString.length > 0 ? inputString[0] : undefined,
    inputIndex: 0,
    remainingInput: inputString,
    actionSummary: `Initial state: ${machine.states.find(s => s.id === currentState)?.label || currentState}`,
    status: 'RUNNING',
  });

  if (!currentState) {
    traces[0].status = 'ERROR';
    traces[0].actionSummary = 'No start state defined for DFA';
    return {
      accepted: false,
      finalStatus: 'ERROR',
      traces,
      totalSteps: 1,
      message: 'No start state defined for DFA.',
    };
  }

  for (let i = 0; i < inputString.length; i++) {
    const symbol = inputString[i];
    
    // Find matching transition
    const transition = machine.transitions.find(
      t => t.from === currentState && t.symbol === symbol
    );

    if (!transition) {
      // Missing transition -> Trap / Dead state -> Immediate reject
      const rejectTrace: SimulationStepTrace = {
        stepIndex: i + 1,
        activeStates: [currentState],
        currentSymbol: symbol,
        inputIndex: i,
        remainingInput: inputString.slice(i),
        actionSummary: `No transition from state ${machine.states.find(s => s.id === currentState)?.label || currentState} on symbol '${symbol}' (Halted & Rejected)`,
        status: 'REJECTED',
      };
      traces.push(rejectTrace);

      return {
        accepted: false,
        finalStatus: 'REJECTED',
        traces,
        totalSteps: traces.length,
        message: `DFA halted: No valid transition for '${symbol}' from state ${currentState}.`,
      };
    }

    currentState = transition.to;
    const nextSymbol = i + 1 < inputString.length ? inputString[i + 1] : undefined;
    const remaining = inputString.slice(i + 1);

    traces.push({
      stepIndex: i + 1,
      activeStates: [currentState],
      currentSymbol: nextSymbol,
      inputIndex: i + 1,
      remainingInput: remaining,
      activeTransitionId: transition.id,
      actionSummary: `Read '${symbol}' → Transitioned to ${machine.states.find(s => s.id === currentState)?.label || currentState}`,
      status: 'RUNNING',
    });
  }

  // End of input reached
  const isAccept = machine.acceptStates.includes(currentState);
  const finalStatus = isAccept ? 'ACCEPTED' : 'REJECTED';
  
  if (traces.length > 0) {
    traces[traces.length - 1].status = finalStatus;
    const lastLabel = machine.states.find(s => s.id === currentState)?.label || currentState;
    traces[traces.length - 1].actionSummary += ` | Input exhausted at ${lastLabel} (${isAccept ? 'ACCEPT' : 'REJECT'})`;
  }

  return {
    accepted: isAccept,
    finalStatus,
    traces,
    totalSteps: traces.length,
    message: isAccept 
      ? `Input string "${inputString}" ACCEPTED by DFA in state ${currentState}.` 
      : `Input string "${inputString}" REJECTED by DFA in non-accepting state ${currentState}.`,
  };
}
